funsor/terms.py

```diff
@@ -994,6 +1001,46 @@ def die_binary(op, lhs, rhs):
     raise NotImplementedError(f"Missing pattern for {repr(expr)}")
 
 
+class Expand(Funsor):
```

Comment:
Hmm I'd like to better understand the need for this.

We've been trying to preserve the [extensionality](https://en.wikipedia.org/wiki/Extensionality) property in Funsor, which states that: if under every grounding substitution `subs` a pair of funsors `f,g` satisfy `f(**subs) == g(**subs)`, then it should be permissible for an optimizer to replace funsor `f` with funsor `g` in any expression.  IIUC this `Expand` funsor would break extensionality because `f.expand(...)` behaves as `f` under every grounding substitution.